x/ibc/23-commitment/types/merkle.go

```diff
@@ -118,34 +128,51 @@ func (mp MerklePath) Pretty() string {
 
 // IsEmpty returns true if the path is empty
 func (mp MerklePath) IsEmpty() bool {
-	return len(mp.KeyPath) == 0
+	return len(mp.KeyPaths) == 0
 }
 
 // ApplyPrefix constructs a new commitment path from the arguments. It interprets
 // the path argument in the context of the prefix argument.
 //
+// ApplyPrefix will prepend a new Merkle KeyPath to the existing KeyPaths. This
+// allows a single path to be represent a chained merkle proof
+//
 // CONTRACT: provided path string MUST be a well formated path. See ICS24 for
 // reference.
-func ApplyPrefix(prefix exported.Prefix, path string) (MerklePath, error) {
-	err := host.DefaultPathValidator(path)
+func ApplyPrefix(prefix exported.Prefix, path exported.Path) (exported.Path, error) {
+	err := host.DefaultPathValidator(path.String())
 	if err != nil {
 		return MerklePath{}, err
 	}
 
 	if prefix == nil || prefix.IsEmpty() {
 		return MerklePath{}, errors.New("prefix can't be empty")
 	}
-	return NewMerklePath([]string{string(prefix.Bytes()), path}), nil
+	mpath, ok := path.(MerklePath)
+	if !ok {
+		return MerklePath{}, errors.New("path is not a merklepath")
+	}
+	// Turn prefix into a merklepath
+	prefixPath := NewMerklePath([]string{string(prefix.Bytes())})
+	kpath := append(prefixPath.KeyPaths, mpath.KeyPaths...)
+	// Return MerklePath with prefix path prepended to list of paths
+	return MerklePath{
+		KeyPaths: kpath,
+	}, nil
 }
 
 var _ exported.Proof = MerkleProof{}
 
-// MerkleProof is a wrapper type that contains a merkle proof.
+// MerkleProof is a wrapper type that contains list of merkle proofs that can
+// be chained to prove membership or non-membership of an element up to the final root.
+// The list is ordered from lowest subtree to the final tree. Each proof's root is used
+// as the proving value for the subsequent proof
 // It demonstrates membership or non-membership for an element or set of elements,
 // verifiable in conjunction with a known commitment root. Proofs should be
 // succinct.
 type MerkleProof struct {
-	Proof *merkle.Proof `json:"proof" yaml:"proof"`
+	Proofs []*ics23.CommitmentProof `json:"proofs" yaml:"proofs"`
```

Comment:
Updated the SDK's MerkleProof to be a list of `CommitmentProof` and specs that are chained from smallest subtree to the top tree, each subroot serving as the value for the next proof.

This allows  clients to define arbitrary chaining of proofs with different proofspecs in their state whether they are TM/SDK chains or not. Also doesn't require putting this chaining logic in confio package

Reply:
I like this approach.